### PR TITLE
Export desktop recordings to video

### PR DESCRIPTION
**Pull request by zmb3**
Add a new tsh command that will write Windows desktop recordings
to an AVI file for offline playback. Encoding is done client side
to avoid consuming server resources.

This uses the [Motion JPEG codec](https://en.wikipedia.org/wiki/Motion_JPEG) for
its simplicity and ease of use. Something like ffmpeg would perform
better in nearly every aspect (run time, compression / file size, video
quality, etc), but that would complicate our build process and add extra
native dependencies. This implementation uses pure Go and works on any
platform where tsh runs today.

Also make sure `tsh recordings ls` shows Windows and SSH recordings.

### CODE REVIEW

File: tool/tsh/recording_export.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+						),
+						fragment,
+						fragment.Bounds().Min,
+						draw.Src,
+					)
+
+					// emit a frame if there's been enough of a time lapse between last event
+					delta := evt.DelayMilliseconds - lastDelay
+					framesToEmit := int64(float64(delta) / frameDelayMillis)
+					lastDelay = evt.DelayMilliseconds
```
Comment (zmb3) [inline tool/tsh/recording_export.go:131]: This technically shouldn't be updated every time like this. `lastDelay` should be renamed to something like `lastEmitted` and only updated when we emit.

**Comment by zmb3** [comment]
Currently untangling a cyclic import with the events tests, but the code can be reviewed.

**Review by ptgott — APPROVED** [review]
The *.mdx pages LGTM

**Review by ibeckermayer — COMMENTED** [review]
The results of tinkering around to get things to work for me locally:

https://github.com/gravitational/teleport/pull/24104

File: lib/client/client.go [hunk 2]
```diff
@@ -2073,8 +2073,9 @@ func GetPaginatedSessions(ctx context.Context, fromUTC, toUTC time.Time, pageSiz
 		if remaining := max - len(sessions); remaining < pageSize {
 			pageSize = remaining
 		}
-		nextEvents, eventKey, err := authClient.SearchSessionEvents(fromUTC, toUTC,
-			pageSize, order, prevEventKey, nil /* where condition */, "" /* session ID */)
+		nextEvents, eventKey, err := authClient.SearchEvents(fromUTC, toUTC, apidefaults.Namespace,
```
[the remark this answers is not in the document]
Comment (zmb3) [reply lib/client/client.go:2056]: Yeah, turns out this wasn't necessary - will revert.

File: docs/pages/reference/cli.mdx [hunk 3]
```diff
@@ -447,6 +447,26 @@ $ tsh play c8e1b2c5-322a-4095-89e3-391edfd2da9b.tar
 ```
 </Admonition>
 
+### tsh recordings export
```
[the remark this answers is not in the document]
Comment (zmb3) [reply docs/pages/reference/cli.mdx:451]: I kind of omitted it on purpose, TBH. Feels more future proof.

File: tool/tsh/recording_export.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		// this should never happen based on what we pass at the call site
+		return nil, trace.BadParameter("unsupported TDP message %T", msg)
+	}
+}
+
+func rectFromPNGMessage(msg tdp.Message) image.Rectangle {
+	switch msg := msg.(type) {
+	case tdp.PNG2Frame:
+		return image.Rect(
+			// add one to bottom and right dimenstion, as RDP
```
Comment (nklaassen) [inline tool/tsh/recording_export.go:209]:
```suggestion
			// add one to bottom and right dimension, as RDP
```

**Review by nklaassen — APPROVED** [review]

**Comment by nklaassen** [comment]
@zmb3  flaky test detector caught something, I like this thing. I take back my approval until that is fixed

File: tool/tsh/recording_export.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+	fname := cf.OutFile
+	if fname == "" {
+		fname = cf.SessionID + ".avi"
+	}
+
+	frames, err := writeMovie(cf.Context, authClient, session.ID(cf.SessionID), fname)
+	// there may be a partial file, even if we encountered an error,
+	// so we indicate to the user when we wrote something
+	if frames > 0 {
+		fmt.Printf("wrote recoring to %v\n", fname)
```
Comment (AntonAM) [inline tool/tsh/recording_export.go:67]:
```suggestion
		fmt.Printf("wrote recording to %v\n", fname)
```

File: tool/tsh/recording_export.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	frameCount := 0
+
+	evts, errs := ss.StreamSessionEvents(ctx, sid, 0)
+loop:
+	for {
+		select {
+		case err := <-errs:
+			if movie != nil {
+				movie.Close()
+				os.Remove(fname)
```
Comment (AntonAM) [inline tool/tsh/recording_export.go:90]: Should we be removing the file (and on context below)? Comment above says there will be partial file even if we encounter an error (and I guess it's also useful to have partial file if user ctrl+c it)

**Review by AntonAM — APPROVED** [review]
Nice work!

File: tool/tsh/recording_export.go (hunk 6 above)
Comment (zmb3) [reply tool/tsh/recording_export.go:90]: Yep, great catch. This is a leftover from before I decided that a partial file makes more sense.

File: lib/events/eventstest/mock_auditlog.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	}()
+
+	return events, errors
+}
+
+func (m *MockAuditLog) EmitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
+	return m.Emitter.EmitAuditEvent(ctx, event)
+}
+
+/*
```
Comment (zmb3) [inline lib/events/eventstest/mock_auditlog.go:57]: Reminder to self to remove this.

**Review by ibeckermayer — APPROVED** [review]

**Pull request merged**